relayer/relays/beacon/cache/cache.go

```diff
@@ -35,6 +68,89 @@ func (b *BeaconCache) SetLastSyncedSyncCommitteePeriod(period uint64) {
 	}
 }
 
+func (b *BeaconCache) AddCheckPoint(finalizedHeaderRoot common.Hash, blockRootsTree *ssz.Node, slot uint64) {
+	b.addSlot(slot)
+	b.Finalized.Checkpoints.Proofs[slot] = Proof{
+		FinalizedBlockRoot: finalizedHeaderRoot,
+		BlockRootsTree:     blockRootsTree,
+		Slot:               slot,
+	}
+
+	b.pruneOldCheckpoints()
+}
+
+func (b *BeaconCache) AddCheckPointSlots(slots []uint64) {
+	for _, slot := range slots {
+		b.addSlot(slot)
+	}
+}
+
+func (b *BeaconCache) GetClosestCheckpoint(slot uint64) (Proof, error) {
+	checkpointSlot, err := b.calculateClosestCheckpointSlot(slot)
+	if err != nil {
+		return Proof{}, err
+	}
+
+	val, ok := b.Finalized.Checkpoints.Proofs[checkpointSlot]
+	if !ok {
+		return Proof{Slot: checkpointSlot}, FinalizedCheckPointNotPopulated
+	}
+
+	return val, nil
+}
+
 func (b *BeaconCache) LastFinalizedHeader() common.Hash {
-	return b.Finalized.Headers[len(b.Finalized.Headers)-1]
+	return b.Finalized.LastSyncedHash
+}
+
+func (b *BeaconCache) pruneOldCheckpoints() {
+	checkpointsCount := len(b.Finalized.Checkpoints.Slots)
+	if checkpointsCount <= FinalizedCheckpointsLimit {
+		return
+	}
+
+	slotsToKeep := b.Finalized.Checkpoints.Slots[checkpointsCount-FinalizedCheckpointsLimit : checkpointsCount]
+	slotsToPrune := b.Finalized.Checkpoints.Slots[0 : checkpointsCount-FinalizedCheckpointsLimit]
+
+	for _, slot := range slotsToPrune {
+		delete(b.Finalized.Checkpoints.Proofs, slot)
+	}
+
+	log.WithField("prunedSlots", slotsToPrune).Info("pruned finalized checkpoint slots from cache")
+
+	b.Finalized.Checkpoints.Slots = slotsToKeep
+}
+
+func (b *BeaconCache) addSlot(slot uint64) {
+	addedAlready := false
+	for _, i := range b.Finalized.Checkpoints.Slots {
+		if i == slot {
+			addedAlready = true
+			break
+		}
+	}
+	if !addedAlready {
+		b.Finalized.Checkpoints.Slots = append(b.Finalized.Checkpoints.Slots, slot)
+	}
+	sort.Slice(b.Finalized.Checkpoints.Slots, func(i, j int) bool { return b.Finalized.Checkpoints.Slots[i] < b.Finalized.Checkpoints.Slots[j] })
+}
+
+func (b *BeaconCache) calculateClosestCheckpointSlot(slot uint64) (uint64, error) {
+	blockRootThreshold := int(b.slotsInEpoch * b.epochsPerSyncCommitteePeriod)
+	for _, i := range b.Finalized.Checkpoints.Slots {
+		if i < slot {
+			continue
+		}
+
+		if i == slot { // if the slot is at a finalized checkpoint, we don't need to do the ancestry proof
+			return i, nil
+		}
+
+		checkpointSlot := int(i) // convert to int since it can be negative
+		if checkpointSlot-blockRootThreshold < int(slot) {
+			return i, nil
+		}
```

Comment:
is it better replace to `if i < slot+blockRootThreshold` to avoid the int cast?

Reply:
then maybe we can also add another check in light client following  https://github.com/Snowfork/snowbridge/blob/36c99689aab88a82529931212d72a2157dd8954b/parachain/pallets/ethereum-beacon-client/src/lib.rs#L575

```
ensure!(
				block_slot + T::MaxSlotsPerHistoricalRoot::get() >= latest_finalized_header_slot,
				Error::<T>::ExpectedFinalizedHeaderNotStored
			);
```